### ISSUE TITLE
How do I access other resources in message handler.

### ISSUE THREAD

**Issue by r0man** [xsc/kithara#3]
Hello Yannick,

this library looks nice and I would like to try it out. Since
this library is based on component I would like to know what is
the recommended way to get access to other resources, a database
component for example, in the message handler.

I would have thought I could declare other dependencies on the
consumer, and access them in the message handler through the
consumer. Something like this:

```
(defn message-handler
  [consumer {:keys [body]}]
  (write-msg-to-db (:db consumer) body)
  {:status :ack})
```

But from the docs it looks like the message handler is only
called with the message. Could you provide an example how this is
supposed to work?

Thanks, Roman.

**Comment by xsc** [comment]
@r0man Good question! I admit it's a little cumbersome due to the layering of components kithara provides (e.g. the connection component wraps the channel component, which wraps the queue component and the actual consumer will always be at the bottom, thus not easily accessible).

Generally, I'd make the message handler function close over the environment. I'll provide some examples once I'm not on my phone but on something with a real keyboard. 

I think this is a very important topic and I agree that dependencies should be injectable as simple as possible.

In the end, kithara's message handler function is not much different from e.g. Ring request handlers (they also take only one parameter, the HTTP request, and have to do something meaningful with it). So it might be worthwhile for us to check how the problem is usually approached in that domain.

**Comment by r0man** [comment]
@xsc Thanks for the explanation. Looking forward for an example.

**Comment by xsc** [comment]
@r0man I've been thinking about this and there should be some potential here. Namely, I'd like to add a `with-env` wrapper that injects an `:env` key into the message map, i.e.:

``` clojure
(defonce rabbitmq-consumer-with-env
  (-> (fn [{:keys [env] :as message}]
        (write! (:db env) message)
        {:status :ack})
      (k/consumer ...)
      ...
      (k/with-env {:db ...})))
```

This seems like the most flexible approach (and its usage stays optional).

As for a more component/system-idiomatic usage, I'd like to make the component provided by `with-env` very flexible, allowing arbitrary keys in the record that will all be propagated to the `:env` key. So, the following would be possible:

``` clojure
(defonce system
  (component/system-map
    :consumer (-> message-handler
                  (k/consumer ...)
                  ...
                  (k/with-env)
                  (component/using [:db :elastic]))
    :db       (map->DB {...})
    :elastic  (map->ES {...})))
(alter-var-root #'system component/start)
```

Thoughts?

**Comment by r0man** [comment]
@xsc Your 2nd code snippet is what I would have expected to build up the consumer. So yes I like this.
However, I see some issues with the first snippet.
- What happens if the payload itself has an `env` key, it would get overwritten. 
- If you use something like Prismatic Schema to validate messages, you would need to add the env to all the message schemas as well, or at least an entry to ignore it. 

What about this:

``` clojure
(defonce rabbitmq-consumer-with-env
  (-> (fn [{:keys [env message]}]
        (write! (:db env) message)
        {:status :ack})
      (k/consumer ...)
      ...
      (k/with-env {:db ...})))
```

Roman

**Comment by xsc** [comment]
@r0man The format of the message map is outlined [here](https://github.com/xsc/kithara#message-format) - the payload is in `:body` and `:body-raw`, so if it contains `:env`, it'll not leak into the top-level map.

Now, the message map already contains some keys that are mostly irrelevant when it comes to validation (like `:channel`, `:redelivered?`, `:exchange`, `:delivery-tag`, `:consumer-tag`), so I really think that schemas should refer to the payload (or only happen after a preprocessing step, e.g. `(select-keys message [:body :routing-key])`), not the whole message map.

The nice thing about the Ring-like handler function is, that it's possible to write generic wrappers around it, e.g. for validation:

``` clojure
(defn wrap-schema
  [message-handler schema]
  (fn [message]
    (if (s/check schema (select-keys message [:routing-key :body]))
      {:status :nack, :message "schema validation failed."}
      (message-handler message))))
```

Or for splitting up environment and message into two parameters:

``` clojure
(defn wrap-env-handler
  [message-handler]
  (fn [{:keys [env] :as message}]
    (message-handler env (dissoc message :env))))
```

I would really like to keep the message map as flat and simple as possible since it is so easy to transform it into whatever format you want before it reaches your handler.

**Comment by r0man** [comment]
@xsc Ah, ok. Then forget what I said, I was not aware of the message structure. :)

**Comment by xsc** [comment]
@r0man I added `with-env` (currently not released, but in `0.1.2-SNAPSHOT`), but I think my personal recommendation would now be to use a wrapper component to gather dependencies. I extended the README with a [section on that topic](https://github.com/xsc/kithara#dependency-injection-wrapper-component).

For completeness' sake, here's a plain component example (since peripheral's macros might be a matter of taste):

``` clojure
(defrecord Consumer [database]
  com.stuartsierra.component/Lifecycle
  (start [this]
    (let [consumer (-> message-handler
                       (k/consumer ...)
                       ...
                       (k/with-env {:database database}))]
      (assoc this :consumer (com.stuartsierra.component/start consumer))))
    (stop [this]
      (update this :consumer com.stuartsierra.component/stop)))
```

I, personally, tend to use this approach, especially if [peripheral](https://github.com/xsc/peripheral) is available to take care of components-within-components.

**Comment by r0man** [comment]
@xsc Cool, this is exactly what I was looking for. Thanks for adding this. I'll give it a try soon ...
